allennlp/modules/token_embedders/bert_token_embedder.py

```diff
@@ -0,0 +1,111 @@
+(https://github.com/huggingface/pytorch-pretrained-BERT),
+so thanks to them!
+"""
+import logging
+
+import torch
+
+from pytorch_pretrained_bert.modeling import BertModel
+
+from allennlp.modules.token_embedders.token_embedder import TokenEmbedder
+from allennlp.nn import util
+
+logger = logging.getLogger(__name__)
+
+
+class BertEmbedder(TokenEmbedder):
+    """
+    A ``TokenEmbedder`` that produces BERT embeddings for your tokens.
+    Should be paired with a ``BertIndexer``, which produces wordpiece ids.
+
+    Most likely you probably want to use ``PretrainedBertEmbedder``
+    for one of the named pretrained models, not this base class.
+
+    Parameters
+    ----------
+    bert_model: ``BertModel``
+        The BERT model being wrapped.
+    """
+    def __init__(self,
+                 bert_model: BertModel) -> None:
+        super().__init__()
+        self.bert_model = bert_model
+        self.output_dim = bert_model.config.hidden_size
+
+    def get_output_dim(self) -> int:
+        return self.output_dim
+
+    def forward(self,
+                input_ids: torch.LongTensor,
+                offsets: torch.LongTensor = None,
+                token_type_ids: torch.LongTensor = None) -> torch.Tensor:
+        """
+        Parameters
+        ----------
+        input_ids: ``torch.LongTensor``
+            The wordpiece ids for each input sentence.
+        offsets: ``torch.LongTensor``, optional (default = None)
+            The BERT embeddings are one per wordpiece. However it's possible/likely
+            you might want one per original token. In that case, ``offsets``
+            should represent the indices of the last wordpiece for each original token.
+
+            That is, if you had the sentence "Definitely not", and if the corresponding
+            wordpieces were ["Def", "##in", "##ite", "##ly", "not"], then the input_ids
+            would be 5 wordpiece ids, and the offsets would be [3, 4]. If offsets are
+            provided, the returned tensor will contain the _last_ wordpiece embedding
+            for each token, and (in particular) will contain one embedding per token.
+            If offsets are not provided, the entire tensor of wordpiece embeddings
+            will be returned.
+        token_type_ids: ``torch.LongTensor``, optional (default = None)
+            If an input consists of two sentences (as in the BERT paper),
+            tokens from the first sentence should have type 0 and tokens from
+            the second sentence should have type 1.  If you don't provide this
+            (the default BertIndexer doesn't) then it's assumed to be all 0s.
+        """
+        # pylint: disable=arguments-differ
+        if token_type_ids is None:
+            token_type_ids = torch.zeros_like(input_ids)
+
+        input_mask = (input_ids != 0).long()
+
+        all_encoder_layers, _ = self.bert_model(input_ids, input_mask, token_type_ids)
+        sequence_output = all_encoder_layers[-1]
```

Comment:
The top layer is a particularly bad layer to use for transfer (talk to Nelson about why, or wait for his final internship presentation).  If we're fine-tuning, this is ok, but if we're not, we probably need some kind of scalar mix or something here, instead of just taking the last layer.  This is probably why your performance is so poor for your NER experiment.

Reply:
Table 2 of the original elmo paper suggests otherwise (unless this is specific to transformers)? Perhaps this accounts for ~1 F1 difference, not 20+.

Reply:
Layer 1 is pretty consistently better than high layers across tasks, in Nelson's experiments (with 2 and 4 layer ELMo).  The top layer is specialized for language modeling, and only a few tasks are close enough to benefit from that specialization (turns out NER is one of those).  For transformers, the story is a bit different, but middle layers are still better than the top layer.

Table 2 of the ELMo paper didn't try just using layer 1.

Reply:
I wasn't saying there isn't a difference - just that even with the top layer, it should be there or thereabouts. So that isn't the problem/difference to look into here for the NER model that Joel trained.

Reply:
I missed that slack convo. Nevermind.